src/uniform.js

```diff
@@ -228,6 +228,64 @@ vgl.modelViewUniform = function (name) {
 
 inherit(vgl.modelViewUniform, vgl.uniform);
 
+///////////////////////////////////////////////////////////////////////////////
+/**
+ * Create new instance of class modelViewOriginUniform.
+ *
+ * @param name
+ * @param uniform: a triplet of floats.
+ * @returns {vgl.modelViewUniform}
+ */
+///////////////////////////////////////////////////////////////////////////////
+vgl.modelViewOriginUniform = function (name, origin) {
+  'use strict';
+
+  if (!(this instanceof vgl.modelViewOriginUniform)) {
+    return new vgl.modelViewOriginUniform(name, origin);
+  }
+
+  if (name.length === 0) {
+    name = 'modelViewMatrix';
+  }
+
+  var m_origin = vec3.fromValues(origin[0], origin[1], origin[2]);
+
+  vgl.uniform.call(this, vgl.GL.FLOAT_MAT4, name);
+
+  this.set(mat4.create());
+
+  /////////////////////////////////////////////////////////////////////////////
+  /**
+   * Update the uniform given a render state and shader program.  This offsets
+   * the modelViewMatrix by the origin, and, if the model view should be
+   * aligned, aligns it appropriately.  The alignment must be done after the
+   * origin offset to maintain precision.
+   *
+   * @param {vgl.renderState} renderState
+   * @param {vgl.shaderProgram} program
+   */
+  /////////////////////////////////////////////////////////////////////////////
+  this.update = function (renderState, program) {
+    program = program; /* unused parameter */
+    var view = mat4.create();
+    mat4.translate(view, renderState.m_modelViewMatrix, m_origin);
+    if (renderState.m_modelViewAlignment) {
+      var align = renderState.m_modelViewAlignment;
+      /* view[12] and view[13] are the x and y offsets.  align.round is the
+       * units-per-pixel, and align.dx and .dy are either 0 or half the size of
+       * a unit-per-pixel.  The alignment guarantees that the texels are
+       * aligned with screen pixels. */
+      view[12] = Math.round(view[12] / align.round) * align.round + align.dx;
```

Comment:
For zoom level 4, a tile is 22.5 degrees across, so the tile edges are the x locations of [-180, -157.5, -135, -112.5, -90, ...].  The tiles have 256 pixels across, so a pixel is 0.087890625 degrees across.  This means, that the edges of the texels are aligned in world x coordinates at locations such as [..., -135, -134.912109375, -134.82421875, -134.736328125, ...].

Our view matrix has no rotation and only an offset, and is aligned with the center of the view port.  Call the camera center `cx`.  view[12] of the view matrix will be `-cx` unless we adjust it.  If the map is `w` pixels across, then the left edge of the map is at `view[12] - w / 2 * 0.087890625`.  We want this to be aligned to the texel edge, which is what the alignment function does.

The projection matrix scales from the zoom level to the screen coordinates.  For my window currently, the projection matrix is [2.5617473125457764, 0, 0, 0, 0, 3.732050895690918, 0, 0, 0, 0, -80.19988250732422, -1, 0, 0, -13592.814453125, 0], and I have a window size of 1502 x 1031 pixels.

You can do the matrix multiplication and see that this works out.